src/api/app/assets/javascripts/webui/application/kiwi_editor.js

```diff
@@ -66,11 +66,41 @@ function closeDialog() {
     }
   }
 
+  addDefault(dialog);
+
   fields.find(".ui-state-error").addClass('hidden');
   $('.overlay').hide();
   dialog.addClass('hidden');
 }
 
+function revertDialog() {
+  var fields = $(this).parents('.nested-fields');
+  var dialog = fields.find('.dialog');
+
+  $.each(dialog.find('input:visible, select:visible'), function(index, input) {
+    if (input.type === 'checkbox') {
+      $(input).prop('checked', input.getAttribute('data-default') === 'true');
+    }
+    else {
+      $(input).val(input.getAttribute('data-default'));
+    }
+  });
+
+  $('.overlay').hide();
+  dialog.addClass('hidden');
+}
+
+function addDefault(dialog) {
```

Comment:
@DavidKang can you explain why we need this method? I don't understand it.. The data-default attributes are already set in the haml views, why do we need to re-set them here in the javascript?

Reply:
@evanrofe, because it's to make multiple modifications at different times. So we will always revert to the last valid data when you press cancel like we talk. The naming of 'data-default' is confusing, maybe we should change it with a better naming.